### ISSUE TITLE
self.copy with symlinks=True does not copy symlink if the .conan directory is a symlink

### ISSUE THREAD

**Issue by danielroedl** [conan-io/conan#5114]
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

To change quickly my conan user home I link the .conan folder of my project folder to ~/.conan.

If I do so and run a `conan create . demo/testing` in a module with a `self.copy("*", symlinks=True)` in the package step the symlinks are not copied.

I look in the sources of conan and find the failing part in the file `conans/client/file_copier.py`. There is a check if the symlink goes out the src folder. Because of using some conversion from relpath to realpath and back the false behaviour happens.

My configuration was like this:
```bash
$ realpath /home/roedl_d1/.conan
/media/big_hdd/git/conan-testing/.conan
```

I can fix this by adding a realpath convertion also for the src argument:
```diff
diff --git a/conans/client/file_copier.py b/conans/client/file_copier.py
index 6913f19e..a74d743c 100644
--- a/conans/client/file_copier.py
+++ b/conans/client/file_copier.py
@@ -161,7 +161,7 @@ class FileCopier(object):
             src_link = os.path.join(src, linked_folder)
             # Discard symlinks that go out of the src folder
             abs_path = os.path.realpath(src_link)
-            relpath = os.path.relpath(abs_path, src)
+            relpath = os.path.relpath(abs_path, os.path.realpath(src))
             if relpath.startswith("."):
                 continue
```

Can you take this change into one of the next releases? Should I create a pull request?

Conan version: 1.14.4
Operating system: Ubuntu 18.10

**Comment by memsharded** [comment]
Hi @danielroedl 

There is one thing I didn't understand:

> To change quickly my conan user home I link the .conan folder of my project folder to ~/.conan.

What do you mean the .conan folder of your project? Projects don't have such a folder, the cache is per user. So if you have a different one for each project it means you defined a CONAN_USER_HOME, but then you linked back that CONAN_USER_HOME to the original default user cache ``~/.conan``? What I am missing?

**Comment by danielroedl** [comment]
I firstly change the `CONAN_USER_HOME` variable each time I work in my project. At this state I have no folder `~/.conan`. Only a folder `~/my_project/.conan` exists because I set `CONAN_USER_HOME=~/my_project` every time I use conan.

Because I don't want to call the `CONAN_USER_HOME=~/my_project` command in every terminal (and because I forgot it from time to time and wonder why the cache is empty) I create a symlink `ln -s ~/my_project/.conan ~/.conan`.

The advantage of this method is that I the cache directory is only changing if I change the symlink. And secondly every terminal uses the same and correct cache directory. When I switch the link target, I also don't have to set the `CONAN_USER_HOME` variable in every terminal which is already open.

**Comment by memsharded** [comment]
I am sorry, still don't fully get it 😅 

How many caches do you have in practice?

> Because I don't want to call the CONAN_USER_HOME=~/my_project command in every terminal (and because I forgot it from time to time and wonder why the cache is empty) I create a symlink ln -s ~/my_project/.conan ~/.conan.

If you don't call CONAN_USER_HOME in the terminal, it will go to the default ~/.conan anyway, so not sure why the symlink is necessary. Just don't set the CONAN_USER_HOME env-var, for that project, right? It seems to me like a double negation, I think I am still missing something. I am not saying that we couldn't do that change if it makes sense, but we would need to understand it better first.

In any case, why would you use different caches per-project? For most cases, the biggest advantage of the cache is that it is very efficient to use it for multiple projects, while collisions will not happen, as multiple versions and multiple binary configurations can co-exist in the cache. If you could elaborate a bit further about your setup, that would be useful feedback.

Thanks!

**Comment by danielroedl** [comment]
I am currently evaluate conan for further usage in our company.

In the past we have experiences with very big caches while building for a while. That's why I tested in that direction in which ways I can change the cache directory. Maybe I find in the future that this is really not needed.

But in my opinion the bug is not related to my "use separate caches for separate projects" scenario. But also if I use the cache every time for every project from `~/.conan` the bug will appear if I want to have my conan cache on a separate hdd. The same way as described above I just set a symlink to the second disk. 

Workaround for this could be I mount the hdd directly to `~/.conan`. But this is not possible if I have more data on this hdd than the conan cache. And just because it's possible to have a symlink as cache directory it should be fixed.

**Comment by memsharded** [comment]
> But in my opinion the bug is not related to my "use separate caches for separate projects" scenario. But also if I use the cache every time for every project from ~/.conan the bug will appear if I want to have my conan cache on a separate hdd. The same way as described above I just set a symlink to the second disk.

Yes, yes, I am not saying it is not a bug, and it shouldn't be fixed. But it is important for us to understand the deep causes, and the users use cases, because sometimes these bugs are an indicator of a missing feature instead.

Ok, please, do open a PR with the above code. It would be great if you could also contribute a test for it, please have a look, and tell if possible (if not, we could try to contribute one) and ask for help if necessary. Also, note that our CI builds for several OSs, including Windows too, so if using ``os.symlink`` you might want to ``@unittest.skipIf(platform.system()=="Windows", "symlinks not in Win")`` the test too. Many thanks!

**Comment by danielroedl** [comment]
Thanks for your explanation!
Because conan has the feature to change the cache directory by using CONAN_USER_HOME I think this don't indicate a missing feature.

So I open a PR and try to add also a test.

**Comment by danielroedl** [comment]
One question according the tests. I think the correct place for a new test testing this issue is `conans/test/functional/old/file_copier_test.py`.

How can I run only the tests in this file? I try a bit using `tox` but without success.

**Comment by danielroedl** [comment]
I found the command to run the tests locally and added a test.
I also opened a PR.

**Comment by memsharded** [comment]
Implemented in #5125
